### ISSUE TITLE
Editor crashes when uploading/inserting media

### ISSUE THREAD

**Issue by LuckynaSan** [swissspidy/web-stories-wp#12784]
## Bug Description

The editor crashes when dragging an image file from Chrome's download bar onto an element. Here is the crash report:

TypeError: Cannot read properties of undefined (reading 'id')

    at StoryProvider (https://lucksany.com/wp-content/plugins/web-stories/assets/js/wp-story-editor.js?ver=62b94bdfd4d9753f46d0:21:12251)
    at https://lucksany.com/wp-content/plugins/web-stories/assets/js/6071.js?ver=62b94bdfd4d9753f46d0:1:266057
    at SnackbarProvider (https://lucksany.com/wp-content/plugins/web-stories/assets/js/1873.js?ver=62b94bdfd4d9753f46d0:9:54908)
    at https://lucksany.com/wp-content/plugins/web-stories/assets/js/6071.js?ver=62b94bdfd4d9753f46d0:1:266057
    at HistoryProvider (https://lucksany.com/wp-content/plugins/web-stories/assets/js/wp-story-editor.js?ver=62b94bdfd4d9753f46d0:17:2986813)
    at https://lucksany.com/wp-content/plugins/web-stories/assets/js/6071.js?ver=62b94bdfd4d9753f46d0:1:266057
    at Media3pApiProvider (https://lucksany.com/wp-content/plugins/web-stories/assets/js/wp-story-editor.js?ver=62b94bdfd4d9753f46d0:32:22270)
    at https://lucksany.com/wp-content/plugins/web-stories/assets/js/6071.js?ver=62b94bdfd4d9753f46d0:1:266057
    at APIProvider (https://lucksany.com/wp-content/plugins/web-stories/assets/js/wp-story-editor.js?ver=62b94bdfd4d9753f46d0:17:2983692)
    at https://lucksany.com/wp-content/plugins/web-stories/assets/js/6071.js?ver=62b94bdfd4d9753f46d0:1:266057
    at ConfigProvider (https://lucksany.com/wp-content/plugins/web-stories/assets/js/wp-story-editor.js?ver=62b94bdfd4d9753f46d0:17:2983397)
    at ErrorBoundary (https://lucksany.com/wp-content/plugins/web-stories/assets/js/wp-story-editor.js?ver=62b94bdfd4d9753f46d0:21:16533)
    at Fe (https://lucksany.com/wp-content/plugins/web-stories/assets/js/6071.js?ver=62b94bdfd4d9753f46d0:1:256831)
    at le (https://lucksany.com/wp-content/plugins/web-stories/assets/js/6071.js?ver=62b94bdfd4d9753f46d0:1:253875)
    at FlagsProvider (https://lucksany.com/wp-content/plugins/web-stories/assets/js/3371.js?ver=62b94bdfd4d9753f46d0:1:205038)
    at StoryEditor (https://lucksany.com/wp-content/plugins/web-stories/assets/js/wp-story-editor.js?ver=62b94bdfd4d9753f46d0:36:37777)

## Steps to Reproduce

1. in Chrome, download image from https://www.pexels.com/ 
2. Go to add a new story
3. Add an element
4. Drag file from download bar on to element

## Screenshots

https://user-images.githubusercontent.com/66372350/206223414-7522f4e0-34ab-47d5-b7c8-f4c3a254eef6.mov

## Additional Context

- Plugin Version: 1.28.0
- WordPress Version:
- Operating System: macOS
- Browser: Chrome

**Comment by Theusainside** [comment]
The webstories editor crashed many times again and again. there are 2 options on that window,
1. Reload
2. Report errors
happen today nearly 9 times with the editor.
my website is http://www.Theusainside.com/, my site’s health is good. this is happening consistently. there is no other error showing.

please hepl

**Comment by miina** [comment]
> The webstories editor crashed many times again and again. there are 2 options on that window,
> 
> 1. Reload
> 2. Report errors
>    happen today nearly 9 times with the editor.
>    my website is http://www.Theusainside.com/, my site’s health is good. this is happening consistently. there is no other error showing.
> 
> please hepl

@Theusainside Until this bug gets fixed, please use an alternative way to upload an image to the story, for example:
1. Open the File in the folder in your computer instead and drag the file from there (instead of using the Chrome download bar directly)
2. Alternatively, use the Media library uploader to Upload an image to the editor and once it's uploaded, drag it from the Library.
<img width="364" alt="Screenshot 2022-12-08 at 15 55 59" src="https://user-images.githubusercontent.com/3294597/206543053-0e8ea7c7-01e1-4ece-bbf2-a95eea1142c7.png">

Hope this helps for now.

**Comment by swissspidy** [comment]
AFAICT the crash occurs in `updateElementsByResourceId` where `resource` is undefined. Probably a race condition somewhere where this reducer is used. Also happens when removing audio from video, uploading via drag & drop from computer, inserting via media modal, etc.

Need to track down the root cause

**Comment by swissspidy** [comment]
So far unable to reproduce this on `main`. Last time when it happened to me was on v1.28.0.

**Comment by swissspidy** [comment]
:bulb: To reproduce, first add some other elements to the page, e.g. some stickers.

The issue is with the `elementIs` usage here (and in other reducers!:

https://github.com/GoogleForCreators/web-stories-wp/blob/c60e8a848d580a8d3a90cbb376089c8665374d83/packages/story-editor/src/app/story/useStoryReducer/reducers/updateElementsByResourceId.ts#L57

At this point `element` is a proxied instance of the actual element object (`WritableDraft<Element>`) and accessing `resource` will return `undefined`.

So I think the proper way to solve this is by wrapping any usage in `original()`. Example:

```ts
.filter((element) => {
  const originalEl = original(element) as Element;
  return elementIs.media(originalEl) && originalEl.resource.id === id;
})
```

Docs: https://immerjs.github.io/immer/original/

We use `elementIs` helpers in many many reducers though. Having to wrap all instances in `original()` hurts readability.

@miina @barklund WDYT?

**Comment by adamdunnage** [comment]
One [user in the support forum](https://wordpress.org/support/topic/web-story-editor-crashes-unexpectedly/#post-16292123) has also pointed out that this is happening when using the upload option from **Insert > Media**. Tested this and I am able to reproduce. 

Recording of this in action:
https://www.loom.com/share/03fa89e8b3db4befa610a49e2b526658

**Comment by swissspidy** [comment]
Yeah it's the same issue. The fix in #12841 will cover that too.
